src/r3d.imageio/r3dinput.cpp

```diff
@@ -0,0 +1,441 @@
+    {
+        close();
+        terminate();
+    }
+    const char* format_name(void) const override { return "r3d"; }
+    int supports(string_view feature) const override
+    {
+        return (feature == "ioproxy");
+    }
+    bool open(const std::string& name, ImageSpec& spec) override;
+    bool open(const std::string& name, ImageSpec& spec,
+              const ImageSpec& config) override;
+    bool seek_subimage(int subimage, int miplevel) override;
+    bool read_native_scanline(int subimage, int miplevel, int y, int z,
+                              void* data) override;
+    void read_frame(int pos);
+    int current_subimage(void) const override
+    {
+        lock_guard lock(*this);
+        return m_subimage;
+    }
+    bool close() override;
+
+    const std::string& filename() const { return m_filename; }
+
+    bool seek(int pos);
+    double fps() const;
+    int64_t time_stamp(int pos) const;
+
+private:
+    std::string m_filename;
+    std::unique_ptr<ImageSpec> m_config;  // Saved copy of configuration spec
+    R3DSDK::Clip* m_clip;
+    R3DSDK::VideoDecodeJob m_job;
+    unsigned char* m_image_buffer;
+    size_t m_adjusted;
+    int m_frames;
+    int m_channels;
+    float m_fps;
+    int m_subimage;
+    int64_t m_nsubimages;
+    int m_last_search_pos;
+    int m_last_decoded_pos;
+    bool m_read_frame;
+    int m_next_scanline;
+    // std::vector<uint16_t> m_pixels;
+
+    void initialize();
+    void reset()
+    {
+        DBG std::cout << "R3dInput::reset()\n";
+
+        ioproxy_clear();
+        m_config.reset();
+        m_clip             = nullptr;
+        m_next_scanline    = 0;
+        m_read_frame       = false;
+        m_subimage         = 0;
+        m_last_search_pos  = 0;
+        m_last_decoded_pos = 0;
+        m_image_buffer     = nullptr;
+        m_adjusted         = 0;
+    }
+
+    void close_file() { reset(); }
+    void terminate();
+};
+
+// Export version number and create function symbols
+OIIO_PLUGIN_EXPORTS_BEGIN
+
+OIIO_EXPORT int r3d_imageio_version = OIIO_PLUGIN_VERSION;
+
+OIIO_EXPORT const char*
+r3d_imageio_library_version()
+{
+    return "R3D";
+}
+
+OIIO_EXPORT ImageInput*
+r3d_input_imageio_create()
+{
+    return new R3dInput;
+}
+
+OIIO_EXPORT const char* r3d_input_extensions[] = { "r3d", nullptr };
+
+OIIO_PLUGIN_EXPORTS_END
+
+
+
+void
+R3dInput::initialize()
+{
+    DBG std::cout << "R3dInput::initialize()\n";
+
+    // initialize SDK
+    // R3DSDK::InitializeStatus init_status = R3DSDK::InitializeSdk(".", OPTION_RED_CUDA);
+    R3DSDK::InitializeStatus init_status
+        = R3DSDK::InitializeSdk("/opt/R3DSDKv8_5_1/Redistributable/linux",
+                                OPTION_RED_NONE);
+    if (init_status != R3DSDK::ISInitializeOK) {
+        R3DSDK::FinalizeSdk();
+        DBG std::cout << "Failed to load R3DSDK Lib: " << init_status << "\n";
+        return;
+    }
+
+    DBG std::cout << "SDK VERSION: " << R3DSDK::GetSdkVersion() << "\n";
+#ifdef GPU
```

Comment:
Where is GPU defined? Or is this just a placeholder for later?

Let's definitely get the non-GPU one finalized and merged first, then move on to GPU.

I have some long-incubating Cuda code that I've been sitting on for a while, that I should revive and maybe it's time to have some general Cuda setup utilities for the various parts of OIIO that might be inclined to want/need them.

Reply:
It is a leftover from the previous standalone proof of concept code.